lib/Yare/App/Services/Rebalancing.hs

```diff
@@ -0,0 +1,279 @@
+module Yare.App.Services.Rebalancing
+  ( Error (..)
+  , service
+  ) where
+
+import Yare.Prelude
+
+import Cardano.Api (TxId, TxOut (TxOut), inAnyShelleyBasedEra, setTxIns)
+import Cardano.Api.Ledger.Lens (mkAdaValue)
+import Cardano.Api.Shelley
+  ( AddressInEra
+  , AlonzoEraOnwards (..)
+  , BuildTx
+  , BuildTxWith (..)
+  , ConwayEraOnwards (..)
+  , CtxTx
+  , KeyWitnessInCtx (KeyWitnessForSpending)
+  , LedgerProtocolParameters (..)
+  , Lovelace
+  , ReferenceScript (..)
+  , ShelleyBasedEra
+  , Tx (..)
+  , TxBodyContent
+  , TxInMode (..)
+  , TxInsCollateral (..)
+  , TxOutDatum (TxOutDatumNone)
+  , TxOutValue
+  , UTxO
+  , Witness (KeyWitness)
+  , calculateMinimumUTxO
+  , constructBalancedTx
+  , convert
+  , defaultTxBodyContent
+  , fromShelleyAddrIsSbe
+  , getTxBody
+  , getTxId
+  , lovelaceToTxOutValue
+  , runExcept
+  , selectLovelace
+  , setTxInsCollateral
+  , setTxOuts
+  , setTxProtocolParams
+  )
+import Cardano.Ledger.Coin (Coin (..))
+import Control.Exception (throwIO)
+import Control.Monad.Error.Class (MonadError (..))
+import Control.Monad.Except (Except)
+import Data.List (zipWith3)
+import Data.Map.Strict qualified as Map
+import GHC.IsList qualified as GHC
+import Text.Pretty.Simple (pShow)
+import Yare.Address (AddressWithKey (..), Addresses (externalAddresses))
+import Yare.Address qualified as Address
+import Yare.Address qualified as Addresses
+import Yare.App.Services.Error (TxConstructionError (..))
+import Yare.App.Types (NetworkInfo (..), StorageMode (..))
+import Yare.Storage (StorageMgr (..), overDefaultStorage)
+import Yare.Submitter qualified as Submitter
+import Yare.Util.State (usingMonadState)
+import Yare.Util.Tx.Construction (minAdaValue, mkCardanoApiUtxo, witnessUtxoEntry)
+import Yare.Utxo (Entry (..), Utxo, spendableEntries)
+import Yare.Utxo qualified as Utxo
+
+service
+  ∷ ∀ era state env
+   . ( [Addresses, Submitter.Q, NetworkInfo era, StorageMgr IO state] ∈∈ env
+     , '[Utxo] ∈∈ state
```

Comment:
FYI: its a better default not to use ticks for promoted constructors, see this ticket https://gitlab.haskell.org/ghc/ghc/-/issues/20531